java/algorithm/src/main/java/org/apache/arrow/algorithm/dictionary/HashTableDictionaryEncoder.java

```diff
@@ -0,0 +1,151 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.arrow.algorithm.dictionary;
+
+import java.util.HashMap;
+
+import org.apache.arrow.memory.util.ArrowBufPointer;
+import org.apache.arrow.memory.util.hash.ArrowBufHasher;
+import org.apache.arrow.memory.util.hash.SimpleHasher;
+import org.apache.arrow.vector.BaseIntVector;
+import org.apache.arrow.vector.ElementAddressableVector;
+
+/**
+ * Dictionary encoder based on hash table.
+ * @param <E> encoded vector type.
+ * @param <D> decoded vector type, which is also the dictionary type.
+ */
+public class HashTableDictionaryEncoder<E extends BaseIntVector, D extends ElementAddressableVector> {
```

Comment:
Are generic necessary here, what value do they provide?

Reply:
I think in general, a generic type makes the parameter type fixed in compilation time, which saves some effort at run time to look up the virtual table for the correct implementation.

Reply:
Can you provide a reference? My understanding was these types are [erased](https://docs.oracle.com/javase/tutorial/java/generics/genTypes.html).  The only times that types can be referenced at runtime is a subclass is create which the pattern that [Guice's TypeLiteral uses](https://google.github.io/guice/api-docs/latest/javadoc/index.html?com/google/inject/TypeLiteral.html)

Reply:
@emkornfield 
I am sorry because my previous comment makes no sense. I should have thought carefully before uttering a comment. 

In general, generic has almost no impact on performance due to type erasure, as you have indicated. The only impact (if any) is related to type cast.

The benefit of generic is that it gives the Java compiler a chance to discover potential (type related) errors in the code, and prevent them from happening at runtime proactively.

This is a general problem. What do you think?

Reply:
What type of errors would the compiler detect  using generics that it wouldn't detected if you replaces all references to E with BaseIntVector and D with ElementAddressableVector?

Reply:
Let's give a concrete example from this PR (I am not sure if it is appropriate).

For this PR, the encoder has a constructor like this,

public HashTableDictionaryEncoder(D dictionary)

and a encode method like this,

public void encode(D input, E output)

The generic makes sure that the input parameter to the constructor and the first parameter to the encode method must have the same type. 

If we do not use generic, the constructor should be:

public HashTableDictionaryEncoder(ElementAddressableVector dictionary)

and the encode method should be 

public void encode(ElementAddressableVector input, ValueVector output)

With the above code, it is possible to construct the encoder with a IntVector, and call the encode method with a VarCharVector. The compiler cannot detect this error, but an exception will be thrown at runtime.

Reply:
Thanks.  That is a good example for "D" is there similar one for "E"?

Reply:
Yes. I think there is a similar example for E (I am not sure if it is appropriate). 

Suppose we have an specific IPC specification to encode VarCharVector with an IntVector. 
With generic, we can declare a encoder like this HashTableEncoder<VarCharVector, IntVector>.

If we break the specification above to encode a VarCharVector by a BigIntVector by calling:

BigIntVector out = ...
encoder.encode(in, out);

With generic, this problem will be detected by the java compiler.

However, without generic, this problem will only be detected at runtime, because BigIntVector is also a sub-type oc BaseIntVector.

Reply:
@emkornfield 
BTW, do you think we should support ListVector/MapVector, etc. with generic types?